### PR TITLE
Strict types, Variants, Options and Product population

### PR DESCRIPTION
**Pull request by psihius**
Implemented strict_types across the plugin.
Added Product variants and options to the elastic index
Updated ProductDocumentFactory to create Variants and Options and push those to Elastic.

### CODE REVIEW

File: src/Controller/AttributeView.php [hunk 1]
```diff
@@ -1,4 +1,5 @@
 <?php
+declare(strict_types=1);
```
Comment (lchrusciel) [inline src/Controller/AttributeView.php]: There should be blank line between `php` tag and `declare`

File: src/Document/OptionDocument.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * @return string
+     */
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    /**
+     * @param string $name
```
Comment (lchrusciel) [inline src/Document/OptionDocument.php:80]: Not needed anymore

Comment (psihius) [reply src/Document/OptionDocument.php:80]: Well, they are all over the code anyway, why leave them out.
Also it makes code beautiful :)

File: src/Document/AttributeDocument.php [hunk 3]
```diff
@@ -53,47 +55,47 @@ class AttributeDocument
     /**
      * @return string
      */
-    public function getCode()
+    public function getCode(): string
```
Comment (pamil) [inline src/Document/AttributeDocument.php:58]: I'm wondering if the string should be nullable. De facto, you can call `getCode` before `setCode`, which will now raise `TypeError`.

File: src/Document/OptionDocument.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+namespace Sylius\ElasticSearchPlugin\Document;
+
+use ONGR\ElasticsearchBundle\Annotation as ElasticSearch;
+
+/**
+ * @ElasticSearch\Nested
+ */
+class OptionDocument
+{
+
```
Comment (pamil) [inline src/Document/OptionDocument.php]: This blank line disturbs me :D

File: src/Document/AttributeDocument.php [hunk 5]
```diff
@@ -53,47 +55,47 @@ class AttributeDocument
...
     {
         return $this->code;
     }
 
     /**
      * @param string $code
      */
-    public function setCode($code)
+    public function setCode(string $code)
```
Comment (pamil) [inline src/Document/AttributeDocument.php]: We can add `void` return types for setters.

File: src/Document/OptionDocument.php (hunk 2 above)
Comment (pamil) [reply src/Document/OptionDocument.php:80]: I'd rather delete them, but it's not a big deal and we can make it package-wide later.

File: src/Document/TaxonDocument.php [hunk 6]
```diff
@@ -29,7 +31,7 @@ class TaxonDocument
      *
      * @ElasticSearch\Property(type="integer")
      */
-    private $position;
+    private $position = 0;
```
Comment (pamil) [inline src/Document/TaxonDocument.php:34]: I'm not sure about this one, why would we need it?

File: src/Document/VariantDocument.php [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * @var Collection
+     *
+     * @ElasticSearch\Embedded(class="Sylius\ElasticSearchPlugin\Document\OptionDocument", multiple=true)
+     */
+    private $options;
+
+    /**
+     * VariantDocument constructor.
```
Comment (pamil) [inline src/Document/VariantDocument.php]: 🎉

File: src/Document/VariantDocument.php [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+     */
+    public function setCode(string $code)
+    {
+        $this->code = $code;
+    }
+
+    /**
+     * @return int
+     */
+    public function getOnHand(): int
```
Comment (pamil) [inline src/Document/VariantDocument.php]: Just a follow-up to our conversation on Slack to replace `onHand` / `onHold` with `stock`.

File: src/Factory/ProductDocumentFactory.php [hunk 9]
```diff
@@ -72,40 +86,57 @@ public function __construct(
 
         $this->assertClassExtends($taxonDocumentClass, TaxonDocument::class);
         $this->taxonDocumentClass = $taxonDocumentClass;
+        $this->priceDocumentClass = $priceDocumentClass;
```
Comment (pamil) [inline src/Factory/ProductDocumentFactory.php]: It looks doubled.

File: src/Factory/ProductDocumentFactory.php [hunk 10]
```diff
@@ -72,40 +86,57 @@ public function __construct(
...
         /** @var ProductTranslationInterface|TranslationInterface $productTranslation */
-        $productTranslation = $syliusProduct->getTranslation($locale->getCode());
-        $channelPrice = $this->getChannelPricingForChannelFromProductVariant($productVariant, $channel);
+        $productTranslation      = $syliusProduct->getTranslation($locale->getCode());
         $syliusProductAttributes = $syliusProduct->getAttributesByLocale(
             $locale->getCode(),
             $channel->getDefaultLocale()->getCode()
         );
-        $syliusProductTaxons = $syliusProduct->getProductTaxons();
+        $syliusProductTaxons     = $syliusProduct->getProductTaxons();
```
Comment (pamil) [inline src/Factory/ProductDocumentFactory.php]: Don't align `=` please, it is a pain to add a new, longer variable then :)

File: src/Factory/ProductDocumentFactory.php [hunk 11]
```diff
@@ -200,7 +237,7 @@ private function createTaxonDocumentFromSyliusTaxon(TaxonInterface $taxon, $loca
         $taxonDocument->setDescription($taxonTranslation->getDescription());
 
         $taxonDocumentImages = [];
-        $syliusTaxonImages = $taxon->getImages();
+        $syliusTaxonImages   = $taxon->getImages();
```
Comment (pamil) [inline src/Factory/ProductDocumentFactory.php]: :(

File: src/Factory/ProductDocumentFactory.php [hunk 12]
```diff
@@ -13,14 +14,17 @@
...
 use Sylius\ElasticSearchPlugin\Document\AttributeDocument;
 use Sylius\ElasticSearchPlugin\Document\ImageDocument;
+use Sylius\ElasticSearchPlugin\Document\OptionDocument;
 use Sylius\ElasticSearchPlugin\Document\PriceDocument;
 use Sylius\ElasticSearchPlugin\Document\ProductDocument;
 use Sylius\ElasticSearchPlugin\Document\TaxonDocument;
-use Sylius\ElasticSearchPlugin\Exception\UnsupportedFactoryMethodException;
+use Sylius\ElasticSearchPlugin\Document\VariantDocument;
 
 final class ProductDocumentFactory implements ProductDocumentFactoryInterface
```
Comment (pamil) [inline src/Factory/ProductDocumentFactory.php:31]: This one needs some refactor, but we can leave that for an another PR.

File: src/Filter/Widget/MultiDynamicAggregateOverride.php [hunk 13]
```diff
@@ -0,0 +1,71 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (pamil) [inline src/Filter/Widget/MultiDynamicAggregateOverride.php]: Settings > Editor > File and Code Templates :)

File: src/Filter/Widget/MultiDynamicAggregateOverride.php [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+use ONGR\FilterManagerBundle\Filter\Widget\Dynamic\MultiDynamicAggregate;
+use ONGR\ElasticsearchDSL\Aggregation\Bucketing\TermsAggregation;
+use ONGR\ElasticsearchDSL\Aggregation\Bucketing\FilterAggregation;
+use ONGR\ElasticsearchDSL\Aggregation\Bucketing\NestedAggregation;
+use ONGR\ElasticsearchDSL\Query\MatchAllQuery;
+use ONGR\ElasticsearchDSL\Search;
+use ONGR\FilterManagerBundle\Filter\FilterState;
+
+class MultiDynamicAggregateOverride extends MultiDynamicAggregate
```
Comment (pamil) [inline src/Filter/Widget/MultiDynamicAggregateOverride.php]: Can you add a comment which explains what does it change?

**Review by pamil — COMMENTED** [review]
Wow, thanks for your work @psihius! Tests need some adjustments though + coverage for the new filters.

File: src/Factory/ProductDocumentFactory.php (hunk 12 above)
Comment (psihius) [reply src/Factory/ProductDocumentFactory.php:31]: You already did some refactoring to this, so we need to merge your and mine  :)

File: src/Factory/ProductDocumentFactory.php (hunk 9 above)
Comment (psihius) [reply src/Factory/ProductDocumentFactory.php]: It is, fixed locally.

File: src/Factory/ProductDocumentFactory.php (hunk 10 above)
Comment (psihius) [reply src/Factory/ProductDocumentFactory.php]: Heh, code auto formatting, i'll change the rules, but I did like how it looked though :)

File: src/Filter/Widget/MultiDynamicAggregateOverride.php (hunk 14 above)
Comment (psihius) [reply src/Filter/Widget/MultiDynamicAggregateOverride.php]: Sure

File: src/Filter/Widget/MultiDynamicAggregateOverride.php (hunk 13 above)
Comment (psihius) [reply src/Filter/Widget/MultiDynamicAggregateOverride.php]: Ok ok :)

File: src/Document/VariantDocument.php (hunk 8 above)
Comment (psihius) [reply src/Document/VariantDocument.php]: Done locally.

File: src/Document/VariantDocument.php (hunk 7 above)
Comment (psihius) [reply src/Document/VariantDocument.php]: Oh yeah, I do! 👍

File: src/Document/TaxonDocument.php (hunk 6 above)
Comment (psihius) [reply src/Document/TaxonDocument.php:34]: If position is not set or is null, this is an integer column and Elastic can't have null values. So we set it by default.

File: src/Document/OptionDocument.php (hunk 4 above)
Comment (psihius) [reply src/Document/OptionDocument.php]: Removed :)

File: src/Document/AttributeDocument.php (hunk 5 above)
Comment (psihius) [reply src/Document/AttributeDocument.php]: Voided the shit out of it

File: src/Document/AttributeDocument.php (hunk 3 above)
Comment (psihius) [reply src/Document/AttributeDocument.php:58]: Well, code is a required field, so it should not return a null. Logically if you don't have a code - attribute should not exist. 

I understand that you can create a document and immediately try to access its code, but I think this should not be possible - I consider that an error in logic. You just don't randomly create an empty object and start to read from it while it's empty.

File: src/Document/OptionDocument.php (hunk 2 above)
Comment (psihius) [reply src/Document/OptionDocument.php:80]: The code style of the whole plugin right now has phpDoc blocks, so I add them to keep it consistent because of broken window theory :)
Just strip them later when it's done for the whole package globally.

**Comment by psihius** [comment]
Well, this ended up being a bit messy, especially because we both changed the ProductDocumentFactory and it turned out into a lot of merge conflicts, so I had to push in all my changes, otherwise this PR was unmergeable.
There are a few things we talked over with @pamil and after this is merged I will create a new PR and this time will make sure I keep it small.

**Comment by psihius** [comment]
Tests are not in a very good place right now and I think we need @Arminek to help us out here a bit :) But I updated shop.yml fixtures - added product variants and product options and product option values to the mix, so we can test those too now.

PR is extending a lot here - adding support for variants, in stock filter, options filter, a few fixes by override for the FilterManagerBundle (there are things that just are plain wrong) and creates a few new filters because they require more than 1 nested level handling.

**Comment by psihius** [comment]
@lchrusciel @pamil ready for review, fixed so tests work. More work will be done on this in future PR's

File: src/Factory/ProductDocumentFactory.php [hunk 15]
```diff
@@ -217,9 +264,55 @@ private function createTaxonDocument(TaxonInterface $taxon, $localeCode = null):
...
+        $variant->setOptions(new Collection($options));
+
+        return $variant;
+    }
+
+    /**
      * @param TaxonInterface $taxon
      *
-     * @return TaxonInterface[]
+     * @return array
```
Comment (pamil) [inline src/Factory/ProductDocumentFactory.php:315]: Actually it's still `TaxonInterface[]`.

File: src/Factory/ProductDocumentFactory.php [hunk 16]
```diff
@@ -283,10 +381,92 @@ private function getChannelPricingForChannelFromProductVariant(
...
+     */
+    private function createAttributeDocumentFromSyliusProductAttributeValue(
+        LocaleInterface $locale,
+        ProductAttributeValueInterface $syliusProductAttributeValue
+    ): array {
+        $productAttributes = [];
+
+        if (is_array($syliusProductAttributeValue->getValue())) {
+            foreach ($syliusProductAttributeValue->getValue() as $value) {
+                $productAttributes[] = $this->createSingleAttributeDocumentFromSyliusProductAttributeValue(
```
Comment (pamil) [inline src/Factory/ProductDocumentFactory.php:428]: Will it actually work?

**Review by pamil — APPROVED** [review]

**Comment by pamil** [comment]
Thanks @psihius, amazing work! 🎉
